pkg/cachemanager/cachemanager.go

```diff
@@ -138,44 +158,84 @@ func (c *CacheManager) UpsertSource(ctx context.Context, sourceKey aggregator.Ke
+		err = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
 	})
 
-	return innerError
+	if err != nil {
+		// account for any watches failing to remove
+		if f := watch.NewErrorList(); errors.As(err, &f) && !f.HasGeneralErr() {
+			removeGVKFailures := watch.SetFrom(f.RemoveGVKFailures())
+			finallyRemoved := c.danglingWatches.Difference(removeGVKFailures)
+
+			c.gvksToDeleteFromCache.AddSet(finallyRemoved)
+			c.danglingWatches.RemoveSet(finallyRemoved)
+			c.danglingWatches.AddSet(removeGVKFailures)
+		} else {
+			// defensively assume all watches that needed removal failed to be removed in the general error case
+			// also assume whatever watches were dangling are still dangling.
+			c.danglingWatches.AddSet(gvksToRemove)
+		}
+
+		return err
+	}
+
+	// if no error, it means no previously dangling watches are still dangling
+	c.gvksToDeleteFromCache.AddSet(c.danglingWatches)
+	c.danglingWatches = watch.NewSet()
+
+	return nil
 }
 
-// removeStaleExpectations stops tracking data for any resources that are no longer watched.
-func (c *CacheManager) removeStaleExpectations(stale *watch.Set) {
-	for _, gvk := range stale.Items() {
-		c.tracker.CancelData(gvk)
+// interpretErr determines if the passed-in error is general (not GVK-specific) and,
+// if GVK-specific, returns the subset of the passed in GVKs that are included in the err.
+func interpretErr(e error, gvks []schema.GroupVersionKind) (bool, []schema.GroupVersionKind) {
+	if e == nil {
+		return false, nil
 	}
+
+	f := watch.NewErrorList()
+	if !errors.As(e, &f) || f.HasGeneralErr() {
+		return true, nil
+	}
+
+	failedGvksToAdd := watch.NewSet()
+	failedGvksToAdd.Add(f.AddGVKFailures()...)
+	sourceGVKSet := watch.NewSet()
+	sourceGVKSet.Add(gvks...)
+
+	common := failedGvksToAdd.Intersection(sourceGVKSet)
+	if common.Size() > 0 {
+		return false, common.Items()
+	}
+
+	// this error is not about the gvks in this request
+	// but we still log it for visibility
+	log.V(logging.DebugLevel).Info("encountered unrelated error when replacing watch set", "error", e)
+	return false, nil
 }
 
 // RemoveSource removes the watches of the GVKs for a given aggregator.Key. Callers are responsible for retrying on error.
 func (c *CacheManager) RemoveSource(ctx context.Context, sourceKey aggregator.Key) error {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	if err := c.gvksToSync.Remove(sourceKey); err != nil {
-		return fmt.Errorf("internal error removing source: %w", err)
-	}
-
-	if err := c.replaceWatchSet(ctx); err != nil {
-		return fmt.Errorf("error removing watches for source %v: %w", sourceKey, err)
+	c.gvksToSync.Remove(sourceKey)
+	err := c.replaceWatchSet(ctx)
+	// Retrying watch deletion due to per-GVK errors is done in the background management loop,
+	// and thus only a general error should be returned to the caller for a controller-based retry.
+	if general, _ := interpretErr(err, []schema.GroupVersionKind{}); general {
+		return fmt.Errorf("error establishing watches: %w", err)
```

Comment:
do we care which sourceKey this error impacted?

Reply:
as in include the the source key in the err here? I think that may be included already in the bubbling up to controller

Reply:
Maybe I'm not looking at the right place, but https://github.com/open-policy-agent/gatekeeper/pull/3030/files/a216503ce3f223be724cc7223f1691fe7b160b81#diff-7b6afcf72ced5f06c1b76b3deff54e9f66c294c5dd8ae7ddf55cdff4eaa3b762R127 is not logging the sync source resulted in the err.

Reply:
I'm not opposed to specifying the key in cache manager or syncset instance in the controller to give more precise indication but here's what it looks like today in the controller logs if there is an error in cachemanager:

```
on remove

Reconciler error   {"controller": "syncset-controller", "object": {"name":"syncset1"}, "namespace": "", "name": "syncset1", "reconcileID": "ee366165-5ad6-44e2-b629-b00d55782e32", "error": "syncset-controller: error removing source: error establishing watches:

on upsert

ERROR    Reconciler error    {"controller": "syncset-controller", "object": {"name":"syncset1"}, "namespace": "", "name": "syncset1", "reconcileID": "e32a25e6-5342-437f-885e-2188041dacdf", "error": "syncset-controller: error upserting watches: error establishing watches:
```

Reply:
ah forgot controller-runtime includes the request object as part of the reconciler error.